articles/130_ros_time.md

```diff
@@ -39,85 +45,110 @@ Slower than real time simulation is necessary for complicated systems where accu
 
-### ROS Time Sources
+When the ROS Time source is active `ROSTime` will return the latest value reported by the Time Source plus the elapsed time since that report multiplied by the real time factor.
 
-#### Default Time Source
+The ROS Time broadcast (clock message) needs to include this information:
+- The current time (in nanoseconds, epoch depends upon source and should not be needed)
+- The current real time factor (also called the "multiplier", typically a value of one, but could be zero for paused or negative when going backwards).
+- The system clock time (in nanoseconds since the Unix epoch) when the message was created (or -1 if not available)
 
-To implement the time abstraction the following approach will be used.
+This message cannot be latched; to use the real time factor you will need to grab a frequency count (aka, start a stopwatch) when the message is received.
+Stale data would make this mechanism fail.
+An exception to this is when all nodes have a system clock available and the message contains the system clock stamp.
+This sytem timestamp could be used to determine the staleness of the "current time" in the message.
+Nodes that publish data before they receive the current time would also be problematic.
+It will be helpful to have the clock publisher publish a new message whenever there is a new subscription added to reduce this period of unknown time.
 
-The time abstraction can be published by one source on the `/clock` topic.
-The topic will contain the most up to date time for the ROS system.
-If a publisher exists for the topic, it will override the system time when using the ROS time abstraction.
-If `/clock` is being published, calls to the ROS time abstraction will return the latest time received from the `/clock` topic.
-If time has not been set it will return zero if nothing has been received.
-A time value of zero should be considered an error meaning that time is uninitialized.
+### ROS Time Message Sources
 
-If the time on the clock jumps backwards, a callback handler will be invoked and be required to complete before any calls to the ROS time abstraction report the new time.
-Calls that come in before that must block.
-The developer has the opportunity to register callbacks with the handler to clear any state from their system if necessary before time will be in the past.
+- An included (where?) Steady Clock Publisher.
+- An included (where?) System Clock Publisher.
+- An included (where?) Gazebo Plugin.
+- The ROSBag playback engine.
 
-The frequency of publishing the `/clock` as well as the granularity are not specified as they are application specific.
+#### Default Time Source
 
-##### No Advanced Estimating Clock
+Users can call `Node::now()` for the current time.
+With the assumption of a single Node instance per process, you can also get the current time from a static method: `Time::now()`.
```

Comment:
Unfortunately I don't think this assumption will ever work. We've been talking at OSRF about the skeleton of the code so that others can help implement it (like you 😀), and I think we've concluded that the "ROS Time" class will only be available from the node class. This ensures that it has the resources it needs to get the "broadcast time" if necessary.

I'm working on a new issue on ros2/ros2 to help guide this effort and to collect details about the detailed code design. hopefully I can finish that this weekend. After that you, and anyone else who is interested, can join us in a hangout or something to discuss it.